### PR TITLE
[containerapp] `az containerapp create/update`: Support --customized-keys and clientType in --bind for dev service

### PR DESCRIPTION
**Pull request by Greedygre**
---

This checklist is used to make sure that common guidelines for a pull request are followed.

### Related command

az containerapp create/update

### General Guidelines

- [ ] Have you run `azdev style <YOUR_EXT>` locally? (`pip install azdev` required)
- [ ] Have you run `python scripts/ci/test_index.py -q` locally? (`pip install wheel==0.30.0` required)

For new extensions:

- [ ] My extension description/summary conforms to the [Extension Summary Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/extensions/extension_summary_guidelines.md).

### About Extension Publish

There is a pipeline to automatically build, upload and publish extension wheels.  
Once your pull request is merged into main branch, a new pull request will be created to update `src/index.json` automatically.  
You only need to update the version information in file setup.py and historical information in file HISTORY.rst in your PR but do not modify `src/index.json`.

### CODE REVIEW

**Comment by yonzhan** [comment]
containerapp

File: src/containerapp/azext_containerapp/containerapp_decorator.py [hunk 1]
```diff
@@ -536,8 +537,20 @@ def set_up_update_containerapp_yaml(self, name, file_name):
             tags = yaml_containerapp.get('tags')
             del yaml_containerapp['tags']
 
+        service_binds = safe_get(yaml_containerapp, "properties", "template", "serviceBinds")
+
```
Comment (Juliehzl) [inline src/containerapp/azext_containerapp/containerapp_decorator.py:541]: add comments for why you do the change

Comment (Juliehzl) [inline src/containerapp/azext_containerapp/containerapp_decorator.py:541]: add defaults?

File: src/containerapp/azext_containerapp/_params.py [hunk 2]
```diff
@@ -24,8 +25,9 @@ def load_arguments(self, _):
         c.argument('artifact', help="Local path to the application artifact for building the container image. See the supported artifacts here: https://aka.ms/SourceToCloudSupportedArtifacts.", is_preview=True)
 
     # Springboard
-    with self.argument_context('containerapp create', arg_group='Service Binding') as c:
+    with self.argument_context('containerapp create', arg_group='Service Binding', is_preview=True) as c:
```
Comment (Juliehzl) [inline src/containerapp/azext_containerapp/_params.py:28]: you want to preview the whole argument group?

File: src/containerapp/azext_containerapp/containerapp_decorator.py [hunk 3]
```diff
@@ -601,6 +617,8 @@ def construct_payload(self):
     def validate_arguments(self):
         super().validate_arguments()
         validate_create(self.get_argument_registry_identity(), self.get_argument_registry_pass(), self.get_argument_registry_user(), self.get_argument_registry_server(), self.get_argument_no_wait(), self.get_argument_source(), self.get_argument_artifact(), self.get_argument_repo(), self.get_argument_yaml(), self.get_argument_environment_type())
+        if self.get_argument_service_bindings() and len(self.get_argument_service_bindings()) > 1 and self.get_argument_customized_keys():
+            raise InvalidArgumentValueError("--bind have multiple values, but --customized-keys only can be set when --bind has single value.")
```
Comment (Juliehzl) [inline src/containerapp/azext_containerapp/containerapp_decorator.py:621]: 1. do not have validation for update command?
2. refine the error message: e.g. --customized-keys only can be set when xxx is single

File: src/containerapp/azext_containerapp/containerapp_decorator.py [hunk 4]
```diff
@@ -768,14 +786,113 @@ def set_up_service_binds(self):
...
+                                                                                            self.get_argument_name(),
+                                                                                            self.get_argument_customized_keys())
             self.set_argument_service_connectors_def_list(service_connectors_def_list)
             unique_bindings = check_unique_bindings(self.cmd, service_connectors_def_list, service_bindings_def_list,
                                                     self.get_argument_resource_group_name(), self.get_argument_name())
             if not unique_bindings:
                 raise ValidationError("Binding names across managed and dev services should be unique.")
             safe_set(self.containerapp_def, "properties", "template", "serviceBinds", value=service_bindings_def_list)
 
+    def set_up_create_containerapp_yaml(self, name, file_name):
```
Comment (Juliehzl) [inline src/containerapp/azext_containerapp/containerapp_decorator.py:798]: do you support update with --yaml

File: src/containerapp/azext_containerapp/_utils.py [hunk 5]
```diff
@@ -74,11 +74,14 @@ def process_service(cmd, resource_list, service_name, arg_dict, subscription_id,
...
                     raise ResourceNotFoundError(f"The service '{service_name}' does not exist")
-
-                service_bindings_def_list.append({
+                service_bind = {
                     "serviceId": containerapp_def["id"],
-                    "name": binding_name
-                })
+                    "name": binding_name,
+                    "clientType": arg_dict.get("clientType")
+                }
```
Comment (Juliehzl) [inline src/containerapp/azext_containerapp/_utils.py:81]: Will it is called by update command? If yes, when --client-type not provided, will it is cleared

File: src/containerapp/azext_containerapp/tests/latest/test_containerapp_commands.py [hunk 6]
```diff
@@ -703,6 +703,198 @@ def test_containerapp_up_dapr_e2e(self, resource_group):
...
+
+        # test update with yaml
+        containerapp_yaml_text = f"""
+                                        properties:
+                                          template:
+                                            serviceBinds:
+                                              - serviceId: {redis_resource_id}
+                                                name: redis
+                                                customizedKeys:
+                                                 CACHE_3_ENDPOINT: REDIS_HOST
```
Comment (Juliehzl) [inline src/containerapp/azext_containerapp/tests/latest/test_containerapp_commands.py:790]: add test to clear customizedKeys

File: src/containerapp/azext_containerapp/tests/latest/test_containerapp_commands.py [hunk 7]
```diff
@@ -703,6 +703,198 @@ def test_containerapp_up_dapr_e2e(self, resource_group):
...
+                JMESPathCheck('length(properties.template.serviceBinds)', 2),
+                JMESPathCheck('properties.template.serviceBinds[0].name', redis_ca_name),
+                JMESPathCheck('properties.template.serviceBinds[0].clientType', "dotnet"),
+                JMESPathCheck('properties.template.serviceBinds[0].customizedKeys.CACHE_1_ENDPOINT', "REDIS_HOST"),
+                JMESPathCheck('properties.template.serviceBinds[0].customizedKeys.CACHE_1_PASSWORD', "REDIS_PASSWORD")
+            ])
+        self.cmd('containerapp delete -n {} -g {} --yes'.format(ca_name, resource_group), expect_failure=False)
+
+        self.cmd(
+            'containerapp create -n {} -g  {} --environment {} --bind {},clientType=dotnet,resourcegroup={} --customized-keys CACHE_2_ENDPOINT=REDIS_HOST CACHE_2_PASSWORD=REDIS_PASSWORD'.format(
```
Comment (Juliehzl) [inline src/containerapp/azext_containerapp/tests/latest/test_containerapp_commands.py:884]: add test cased for update commands to clear customized-keys

File: src/containerapp/azext_containerapp/_utils.py (hunk 5 above)
Comment (Greedygre) [reply src/containerapp/azext_containerapp/_utils.py:81]: No, for update command, we get the value from existing containerapp first, if the input `clientType` not exists, we will not set the value requestBody. 
We have test for `test update customizedKeys without clientType`
image

File: src/containerapp/azext_containerapp/containerapp_decorator.py (hunk 4 above)
Comment (Greedygre) [reply src/containerapp/azext_containerapp/containerapp_decorator.py:798]: Yes, we support, customer need to input the whole `serviceBinds`

File: src/containerapp/azext_containerapp/tests/latest/test_containerapp_commands.py (hunk 6 above)
Comment (Greedygre) [reply src/containerapp/azext_containerapp/tests/latest/test_containerapp_commands.py:790]: Done

File: src/containerapp/azext_containerapp/tests/latest/test_containerapp_commands.py (hunk 7 above)
Comment (Greedygre) [reply src/containerapp/azext_containerapp/tests/latest/test_containerapp_commands.py:926]: We can use update with --yaml to clear the `customizedKeys`.

**Review by Juliehzl — APPROVED** [review]

File: src/containerapp/azext_containerapp/_params.py [hunk 8]
```diff
@@ -24,8 +25,9 @@ def load_arguments(self, _):
...
         c.argument('service_bindings', nargs='*', options_list=['--bind'], help="Space separated list of services(bindings) to be connected to this app. e.g. SVC_NAME1[:BIND_NAME1] SVC_NAME2[:BIND_NAME2]...")
+        c.argument('customized_keys', options_list=['--customized-keys'], action=AddCustomizedKeys, nargs='*', help='The customized keys used to change default configuration names. Key is the original name, value is the customized name.')
```
Comment (zhoxing-ms) [inline src/containerapp/azext_containerapp/_params.py:30]: There is no need to explicitly define the default option name

```suggestion
        c.argument('customized_keys', action=AddCustomizedKeys, nargs='*', help='The customized keys used to change default configuration names. Key is the original name, value is the customized name.')
```

File: src/containerapp/azext_containerapp/_params.py [hunk 9]
```diff
@@ -44,8 +46,9 @@ def load_arguments(self, _):
         c.argument('artifact', help="Local path to the application artifact for building the container image. See the supported artifacts here: https://aka.ms/SourceToCloudSupportedArtifacts.", is_preview=True)
 
     # Springboard
-    with self.argument_context('containerapp update', arg_group='Service Binding') as c:
+    with self.argument_context('containerapp update', arg_group='Service Binding', is_preview=True) as c:
         c.argument('service_bindings', nargs='*', options_list=['--bind'], help="Space separated list of services(bindings) to be connected to this app. e.g. SVC_NAME1[:BIND_NAME1] SVC_NAME2[:BIND_NAME2]...")
+        c.argument('customized_keys', options_list=['--customized-keys'], action=AddCustomizedKeys, nargs='*', help='The customized keys used to change default configuration names. Key is the original name, value is the customized name.')
```
Comment (zhoxing-ms) [inline src/containerapp/azext_containerapp/_params.py:51]:
```suggestion
        c.argument('customized_keys', action=AddCustomizedKeys, nargs='*', help='The customized keys used to change default configuration names. Key is the original name, value is the customized name.')
```

**Review by zhoxing-ms — APPROVED** [review]

**Pull request merged**